src/webgpu/shader/validation/tokens.spec.ts

```diff
@@ -0,0 +1,187 @@
+export const description = `Validation tests for tokenization`;
+
+import { makeTestGroup } from '../../../common/framework/test_group.js';
+
+import { ShaderValidationTest } from './shader_validation_test.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+g.test('empty')
+  .desc(`Test that an empty source file is consumed successfully.`)
+  .fn(t => {
+    t.expectCompileResult(true, '');
+  });
+
+g.test('null_characters')
+  .desc(`Test that WGSL source containing a null character is rejected.`)
+  .params(u =>
+    u
+      .combine('contains_null', [true, false])
+      .combine('placement', ['comment', 'delimiter', 'eol'])
+      .beginSubcases()
+  )
+  .fn(t => {
+    let code = '';
+    if (t.params.placement === 'comment') {
+      code = `// Here is a ${t.params.contains_null ? '\0' : 'Z'} character`;
+    } else if (t.params.placement === 'delimiter') {
+      code = `let${t.params.contains_null ? '\0' : ' '}name : i32 = 0;`;
+    } else if (t.params.placement === 'eol') {
+      code = `let name : i32 = 0;${t.params.contains_null ? '\0' : ''}`;
+    }
```

Comment:
@kainino0x or somebody else that understands Javascript: I'm not actually sure whether this is correct or not. Should these JS strings containing null characters get passed through the WebGPU implementation verbatim, or does something else cause them to be null terminated?

Reply:
Not 100% sure if this answers the question, but:

JS strings are not null-terminated, internally they're essentially a `vector<uint16_t>`.
When we get a string in Blink for example, it arrives as WTF::String or something like that, and we can convert to UTF-8, and NULs inside the string get carried through.
(NUL is the 0 codepoint, encoded as 0x0000 in utf-16 and 0x00 in utf-8. The UTF-8 string resulting from that conversion should include any NULs and any text after them, and if it's std::string, there's also an extra NUL past the end for null-termination).

Reply:
Resolved offline - the tests are fine, there's an issue with our stack that we can address later.